src/lamp_py/mssql/mssql_utils.py

```diff
@@ -161,14 +159,26 @@ def write_to_parquet(
             stream_results=True,
             max_row_buffer=batch_size,
         )
-        with self.session.begin() as cursor:
-            with pq.ParquetWriter(write_path, schema=schema) as pq_writer:
-                for part in cursor.execute(part_stmt).partitions(batch_size):
-                    pq_writer.write_batch(
-                        pyarrow.RecordBatch.from_pylist(
-                            [row._asdict() for row in part], schema=schema
-                        )
-                    )
-
-        process_logger.log_complete()
-        # pylint: enable=R0801
+        max_retries = 3
+        for retry_attempts in range(max_retries + 1):
+            process_logger.add_metadata(retry_attempts=retry_attempts)
+            try:
+                with self.session.begin() as cursor:
+                    with pq.ParquetWriter(
+                        write_path, schema=schema
+                    ) as pq_writer:
+                        for part in cursor.execute(part_stmt).partitions(
+                            batch_size
+                        ):
+                            pq_writer.write_batch(
+                                pyarrow.RecordBatch.from_pylist(
+                                    [row._asdict() for row in part],
+                                    schema=schema,
+                                )
+                            )
+
+                process_logger.log_complete()
+
+            except Exception as exception:
+                if retry_attempts == max_retries:
+                    process_logger.log_failure(exception)
```

Comment:
no reason to suspect that the errors on successive retries, right? if they could happen it might be worth logging the error type each time?

Reply:
And do we want to have a backoff with each try before rolling right into the next session?

Reply:
We've only seen 1 error type, thus far... The ProcessLogger class doesn't currently support logging errors that don't end a process, so that logging would be occurring outside of the process_logger. 

For the error we've been seeing, I don't think a backoff would be beneficial, but that could change. The current error is the DB killing the query and we can't really know whether that will or will not happen or if a backoff would help.

Reply:
i think you can just call `log_failure` multiple times. that does have the effect of spamming our alerts though. probably a good future feature for our process logger.